### PR TITLE
Use the correct DSN format in all DB_HOST scenarios

### PR DESCRIPTION
**Pull request by frankiejarrett**
When attempting to use a host with a port (`mysql.example.com:3307`) I encountered a fatal because `DB_HOST` still has the port in it.

```
Fatal error:  Uncaught exception 'PDOException' with message 'SQLSTATE[HY000] [2005] Unknown MySQL server host
```

Proposed changes to fix the above, and more:

1. Include the modified (portless) `$host` in the return array after parsing `DB_HOST`
2. Use the portless `$host` when a `$port` exists
3. Only use the `port` element when a `$port` exists
4. Use the full `DB_HOST` when the host is portless, or on socket connections
5. Support the `charset` element

https://codex.wordpress.org/Editing_wp-config.php#Set_Database_Host
https://secure.php.net/manual/en/ref.pdo-mysql.connection.php

### CODE REVIEW

File: object-cache.php [hunk 1]
```diff
@@ -1148,13 +1148,20 @@ public function __construct() {
...
 			}
 
-			$dsn = 'mysql:host='. DB_HOST. ';port='. $port .';dbname='. DB_NAME;
+			if ( $socket ) {
+				$dsn = sprintf( 'mysql:unix_socket=%s;dbname=%s;charset=%s', DB_HOST, DB_NAME, DB_CHARSET );
+			} elseif ( $port ) {
+				$dsn = sprintf( 'mysql:host=%s;port=%d;dbname=%s;charset=%s', $host, $port, DB_NAME, DB_CHARSET );
+			} else {
+				$dsn = sprintf( 'mysql:host=%s;dbname=%s;charset=%s', DB_HOST, DB_NAME, DB_CHARSET );
+			}
```
Comment (danielbachhuber) [inline object-cache.php]: What do you think about replacing `private function get_port_socket_from_host()` with `public static function get_pdo_dsn( $host, $name, $charset )`, and writing some unit tests for the latter?

Comment (frankiejarrett) [reply object-cache.php]: @danielbachhuber Love that idea.

**Comment by danielbachhuber** [comment]
Working on the failed Travis test in #116

**Comment by danielbachhuber** [comment]
@fjarrett You can merge master now to get the fix for Travis.

File: object-cache.php (hunk 1 above)
Comment (frankiejarrett) [reply object-cache.php]: @danielbachhuber Done in https://github.com/lcache/wp-lcache/pull/115/commits/b5da7055c4a1ec8cd52e768fe43335fb6dda8bcd

**Comment by danielbachhuber** [comment]
@fjarrett This looks good to me. Happy with it?

**Comment by danielbachhuber** [comment]
If you merge master to include #118, Circle CI should pass.

**Comment by danielbachhuber** [comment]
> Circle CI should pass.

Ugh. Terminus changed its behavior.

**Comment by danielbachhuber** [comment]
Thanks for your work on this, @fjarrett !

**Pull request merged**